File: www/frontend/src/api/message.ts
import { api } from './index'

// ── 类型定义 ──────────────────────────────────────────────────────────────────

/** 与项目其他 API 模块保持一致的通用响应包装器 */
export interface ApiResponse<T> {
  code: number
  success: boolean
  message: string
  data: T
}

export interface Message {
  messageId: number
  senderUserId: number | null
  recipientUserId: number
  subject: string | null
  content: string
  messageType: string
  priority: number
  status: 'UNREAD' | 'READ' | 'DELETED'
  sendTime: string
  readTime: string | null
  senderUsername: string | null
  recipientUsername: string | null
}

export interface SendMessageRequest {
  recipientUserId: number
  subject?: string
  content: string
  messageType?: string
  priority?: number
}

export interface PageResult<T> {
  content: T[]
  totalElements: number
  totalPages: number
  page: number
  size: number
  numberOfElements: number
}

// ── API 方法 ──────────────────────────────────────────────────────────────────

export const messageApi = {
  /** 获取收件箱（分页） */
  getInbox(page = 0, size = 20, msgType?: string) {
    return api.get<ApiResponse<PageResult<Message>>>('/api/messages', {
      params: { page, size, msg_type: msgType },
    })
  },

  /** 获取已发送（分页） */
  getSent(page = 0, size = 20) {
    return api.get<ApiResponse<PageResult<Message>>>('/api/messages/sent', {
      params: { page, size },
    })
  },

  /** 获取未读消息数 */
  getUnreadCount() {
    return api.get<ApiResponse<{ count: number }>>('/api/messages/unread/count')
  },

  /** 获取未读消息列表 */
  getUnreadList() {
    return api.get<ApiResponse<Message[]>>('/api/messages/unread')
  },

  /** 发送消息 */
  send(body: SendMessageRequest) {
    return api.post<ApiResponse<Message>>('/api/messages', body)
  },

  /** 标记单条已读 */
  markRead(messageId: number) {
    return api.put<ApiResponse<null>>(`/api/messages/${messageId}/read`)
  },

  /** 全部标记已读 */
  markAllRead() {
    return api.put<ApiResponse<null>>('/api/messages/read-all')
  },

  /** 删除消息（软删除） */
  remove(messageId: number) {
    return api.delete<ApiResponse<null>>(`/api/messages/${messageId}`)
  },

  /** 获取与某用户的对话 */
  getConversation(userId: number, page = 0, size = 20) {
    return api.get<ApiResponse<PageResult<Message>>>(`/api/messages/conversation/${userId}`, {
      params: { page, size },
    })
  },
}

File: www/backend/models/message.py
"""
消息数据模型
"""
from datetime import datetime
from enum import Enum

from sqlalchemy import (
    BigInteger, Column, DateTime, ForeignKey,
    Integer, String, Text,
)
from sqlalchemy.orm import relationship

from database import Base


class MessageStatus(str, Enum):
    UNREAD = "UNREAD"
    READ = "READ"
    DELETED = "DELETED"


class Message(Base):
    """消息实体，映射 `messages` 表"""
    __tablename__ = "messages"

    message_id        = Column(BigInteger, primary_key=True, autoincrement=True)
    sender_user_id    = Column(BigInteger, ForeignKey("user.user_id"), nullable=True)   # None = 系统消息
    recipient_user_id = Column(BigInteger, ForeignKey("user.user_id"), nullable=False)
    subject           = Column(String(200), nullable=True)
    content           = Column(Text, nullable=False)
    message_type      = Column(String(50), nullable=False, default="USER")   # USER / SYSTEM / REVIEW
    priority          = Column(Integer, nullable=False, default=1)
    status            = Column(String(20), nullable=False, default=MessageStatus.UNREAD)
    send_time         = Column(DateTime, nullable=False, default=datetime.now)
    read_time         = Column(DateTime, nullable=True)

    sender    = relationship("User", foreign_keys=[sender_user_id])
    recipient = relationship("User", foreign_keys=[recipient_user_id])

    # ── 快捷属性，供 schema 序列化时使用 ─────────────────────────────────────
    @property
    def sender_username(self) -> str | None:
        return self.sender.username if self.sender else None

    @property
    def recipient_username(self) -> str | None:
        return self.recipient.username if self.recipient else None
